From 94cf29c4e3b6a7f37ad9d85ba572409691ed7f23 Mon Sep 17 00:00:00 2001
From: Luc Maisonobe <luc@apache.org>
Date: Fri, 29 Aug 2008 15:46:22 +0000
Subject: [PATCH] Changed the Complex.equals() method so that it considers +0
 and -0 are equal, as required by IEEE-754 standard. JIRA: MATH-221

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/math/branches/MATH_2_0@690308 13f79535-47bb-0310-9956-ffa450edef68
---
 src/java/org/apache/commons/math/complex/Complex.java     | 5 +----
 src/site/xdoc/changes.xml                                 | 4 ++++
 src/test/org/apache/commons/math/complex/ComplexTest.java | 6 ++++++
 3 files changed, 11 insertions(+), 4 deletions(-)

diff --git a/src/java/org/apache/commons/math/complex/Complex.java b/src/java/org/apache/commons/math/complex/Complex.java
index 16b3873323..8b622cb295 100644
--- a/src/java/org/apache/commons/math/complex/Complex.java
+++ b/src/java/org/apache/commons/math/complex/Complex.java
@@ -255,10 +255,7 @@ public boolean equals(Object other) {
                 if (rhs.isNaN()) {
                     ret = this.isNaN();
                 } else {
-                ret = (Double.doubleToRawLongBits(real) ==
-                        Double.doubleToRawLongBits(rhs.getReal())) &&
-                    (Double.doubleToRawLongBits(imaginary) ==
-                        Double.doubleToRawLongBits(rhs.getImaginary())); 
+                    ret = (real == rhs.real) && (imaginary == rhs.imaginary); 
                 }
             } catch (ClassCastException ex) {
                 // ignore exception
diff --git a/src/site/xdoc/changes.xml b/src/site/xdoc/changes.xml
index ee0980f1f9..a38c39c210 100644
--- a/src/site/xdoc/changes.xml
+++ b/src/site/xdoc/changes.xml
@@ -39,6 +39,10 @@ The <action> type attribute can be add,update,fix,remove.
   </properties>
   <body>
     <release version="2.0" date="TBD" description="TBD">
+      <action dev="luc" type="fix" issue="MATH-221" due-to="Dieter Roth">
+        Changed the Complex.equals() method so that it considers +0 and -0 are equal,
+        as required by IEEE-754 standard.
+      </action>
       <action dev="luc" type="add" issue="MATH-220" >
         Added JAMA-like interfaces for decomposition algorithms. These interfaces
         decompose a matrix as a product of several other matrices with predefined
diff --git a/src/test/org/apache/commons/math/complex/ComplexTest.java b/src/test/org/apache/commons/math/complex/ComplexTest.java
index 4ba7a65134..61ff86f52a 100644
--- a/src/test/org/apache/commons/math/complex/ComplexTest.java
+++ b/src/test/org/apache/commons/math/complex/ComplexTest.java
@@ -690,4 +690,10 @@ public void testTanhInf() {
     public void testTanhCritical() {
         TestUtils.assertSame(nanInf, new Complex(0, pi/2).tanh());
     }
+
+    /** test issue MATH-221 */
+    public void testMath221() {
+        assertEquals(new Complex(0,-1), new Complex(0,1).multiply(new Complex(-1,0)));
+    }
+
 }
